File: ai_interview_service/routers/interview.py
from fastapi import APIRouter, File, Form, HTTPException, UploadFile, status

from models.interview import InteriewStatusEnum
from request_model.AnswerRequest import AnswerRequest
from response_model.AnswerResponse import AnswerResponse
from services.ai_service import create_interview_context, evaluate_interview_answers
from services.interview_service import create_session, get_session, save_answer
from util.file_util import extract_text, validate_file

router = APIRouter(
    prefix="/interview",
    tags=["Interview"]
)

@router.post("/generate-question")
async def init_interview_context(
    job_title: str = Form(...),
    job_description: str = Form(...),
    resume: UploadFile = File(...)
):
    # Validate resume file
    await validate_file(resume)

    # Extract text from resume
    resume_text = await extract_text(resume)

    # Generate questions and intro text using title, decription, resume content
    context_data = await create_interview_context(job_title=job_title, job_description=job_description, resume_text=resume_text)

    session = create_session()
    session.questions = context_data.get("questions")
    session.introText = context_data.get("introText")

    return { "session_id": session.session_id }



@router.get("/start/{session_id}")
async def start_interview(session_id: str):
    # Check valid session_id
    session = get_session(session_id)
    if not session or session.status == InteriewStatusEnum.COMPLETED:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Interview Session Not Found")

    return {
        "introText": session.introText,
        "firstQuestion": session.questions[0]
    }


@router.post("/submit", response_model=AnswerResponse)
async def submit_answer(answerReq: AnswerRequest):
    # Check valid session_id
    session = get_session(answerReq.session_id)
    if not session or session.status == InteriewStatusEnum.COMPLETED:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Interview Session Not Found")


    # save answer in interview session
    save_answer(answerReq.answer, answerReq.skip, session)

    # return
    if session.status == InteriewStatusEnum.COMPLETED:
        return {
            "interviewEnded": True
        }
    
    return {
        "interviewEnded": False,
        "nextQuestion": session.questions[session.current_index]
    }

@router.put("/end/{session_id}")
async def end_interview(session_id: str):
    # Check valid session_id
    session = get_session(session_id)
    if not session or session.status == InteriewStatusEnum.COMPLETED:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Interview Session Not Found")
    
    session.status = InteriewStatusEnum.COMPLETED

    return {
            "interviewEnded": True
    }

@router.get("/report/{session_id}")
async def get_interview_report(session_id: str):
    # Check valid session_id
    session = get_session(session_id)
    if not session:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Interview Session Not Found")
    
    evaluation_data = await evaluate_interview_answers(answers=session.answers)

    return { "result":  evaluation_data }

File: ai_interview_frontend/src/services/interview.js
import axios from "axios";

const BASE_URL = "http://localhost:8000/interview";

export const startInterviewAPI = async (sessionId) => {
  try {
    const response = await axios.get(`${BASE_URL}/start/${sessionId}`);
    return response.data;
  } catch (error) {
    console.error("Error: ", error);
  }
  return null;
};

export const submitAPI = async (payload) => {
  try {
    const response = await axios.post(`${BASE_URL}/submit`, payload);
    return response.data;
  } catch (error) {
    console.error("Error: ", error);
  }
  return null;
};

export const reportAPI = async (sessionId) => {
  try {
    const response = await axios.get(`${BASE_URL}/report/${sessionId}`);
    return response.data;
  } catch (error) {
    console.error("Error: ", error);
  }
  return null;
};

export const endInterviewAPI = async (sessionId) => {
  try {
    const response = await axios.put(`${BASE_URL}/end/${sessionId}`);
    return response.data;
  } catch (error) {
    console.error("Error: ", error);
  }
  return null;
};

export const generateQuestionsAPI = async (formData) => {
  try {
    const response = await axios.post(
      `${BASE_URL}/generate-question`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data"
        },
      }
    );
    return response.data;
  } catch (error) {
    console.error("Error: ", error);
  }
  return null;
};
